File: db_lease.go
package kvr

import (
	"bytes"
	"encoding/binary"

	"github.com/PowerDNS/lmdb-go/lmdb"
)

type dbLease struct {
	db
}

type lease struct {
	id      uint64
	expires uint64
	renewed uint64
}

func (item lease) FromBytes(key, buf []byte) (lease, error) {
	var err error
	if len(buf) < 6 {
		return item, ErrChecksumMissing
	}
	if binary.BigEndian.Uint32(buf[len(buf)-4:]) != crc(key, buf[:len(buf)-4]) {
		return item, ErrChecksumInvalid
	}
	var n int
	if item.id, n = binary.Uvarint(key); n == 0 {
		return item, ErrLeaseKeyInvalid
	}
	r := bytes.NewBuffer(buf[:len(buf)-4])
	if item.expires, err = binary.ReadUvarint(r); err != nil {
		return item, err
	}
	if item.renewed, err = binary.ReadUvarint(r); err != nil {
		return item, err
	}
	return item, nil
}

func (item lease) Bytes(buf []byte) []byte {
	buf = binary.AppendUvarint(buf, item.expires)
	buf = binary.AppendUvarint(buf, item.renewed)
	buf = binary.BigEndian.AppendUint32(buf, crc(binary.AppendUvarint(nil, item.id), buf))
	return buf
}

func newDbLease(txn *lmdb.Txn) (db dbLease, err error) {
	db.i, err = txn.OpenDBI("lease", uint(lmdb.Create))
	return
}

func (db dbLease) get(txn *lmdb.Txn, id uint64) (item lease, err error) {
	k := binary.AppendUvarint(nil, id)
	v, err := txn.Get(db.i, k)
	if lmdb.IsNotFound(err) {
		err = nil
		return
	}
	if err != nil {
		return
	}
	return item.FromBytes(k, v)
}

func (db dbLease) put(txn *lmdb.Txn, item lease) error {
	return txn.Put(db.i, binary.AppendUvarint(nil, item.id), item.Bytes(nil), 0)
}

func (db dbLease) all(txn *lmdb.Txn) (items []lease, err error) {
	cur, err := txn.OpenCursor(db.i)
	if err != nil {
		return nil, err
	}
	defer cur.Close()
	var item lease
	k, v, err := cur.Get(nil, nil, lmdb.Next)
	for !lmdb.IsNotFound(err) && len(k) > 0 {
		if err != nil {
			return nil, err
		}
		item, err := item.FromBytes(k, v)
		if err != nil {
			return nil, err
		}
		items = append(items, item)
		k, v, err = cur.Get(nil, nil, lmdb.Next)
	}
	if lmdb.IsNotFound(err) {
		err = nil
	}
	return
}

func (db dbLease) del(txn *lmdb.Txn, id uint64) error {
	return txn.Del(db.i, binary.AppendUvarint(nil, id), nil)
}

File: module-zig/storage-kv/db_lease.zig
//! Mirrors module/storage-kv/db_lease.go

const std = @import("std");
const lmdb = @import("lmdb");

const Db = @import("db.zig").Db;
const Lease = @import("lease.zig").Lease;
const util = @import("util.zig");

pub const DbLease = struct {
    db: Db,

    pub fn init(self: DbLease, txn: lmdb.Txn) void {
        self.db.open(txn);
    }

    pub fn get(self: DbLease, txn: lmdb.Txn, id: u64) !Lease {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, id)];
        const v = txn.get(self.db.i, k) catch |err| {
            if (err == lmdb.Error.NotFound) {
                return Lease{};
            }
            return err;
        };
        return Lease.fromBytes(k, v);
    }

    pub fn put(self: DbLease, txn: lmdb.Txn, item: Lease) !void {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, item.id)];
        var vbuf: [2 * util.max_varint_len + 4]u8 = undefined;
        return txn.put(self.db.i, k, item.bytes(&vbuf), 0);
    }

    pub fn all(self: DbLease, txn: lmdb.Txn, allocator: std.mem.Allocator) ![]Lease {
        var items = std.ArrayList(Lease).empty;
        const cur = try txn.openCursor(self.db.i);
        defer cur.close();
        while (true) {
            const entry = cur.get("", "", lmdb.op_next) catch |err| {
                if (err == lmdb.Error.NotFound) break;
                return err;
            };
            if (entry.key.len == 0) break;
            const item = try Lease.fromBytes(entry.key, entry.val);
            try items.append(allocator, item);
        }
        return items.items;
    }

    pub fn del(self: DbLease, txn: lmdb.Txn, id: u64) !void {
        var kbuf: [util.max_varint_len]u8 = undefined;
        const k = kbuf[0..util.putUvarint(&kbuf, id)];
        return txn.del(self.db.i, k, "");
    }
};
